src/opamPrinter.ml

```diff
@@ -644,20 +646,176 @@ module FullPos = struct
+            let acc, lastpos = bef_w_elem acc lastpos id in
+            let acc, lastpos = bef_w_elem acc lastpos op in
+            let acc, lastpos = bef_w_elem acc lastpos v in
+            acc, lastpos
+        in
+        acc, lastpos
+      in
+      let rec nc_item (acc, lastpos) item =
+        let acc, _ = before acc lastpos item in
+        match item.pelem with
+        | Variable (name, value) ->
+          let acc, lastpos = element acc name in
+          nc_value (acc, lastpos) value
+        | Section {section_kind; section_name; section_items} ->
+          let acc, lastpos = element acc section_kind in
+          let acc, lastpos =
+            match section_name with
+            | None -> acc, lastpos
+            | Some name ->
+              let acc, lastpos =
+                bef_w_elem acc lastpos name
+              in
+              acc, lastpos
+          in
+          let acc, lastpos = before acc lastpos section_items in
+          List.fold_left nc_item (acc, lastpos) section_items.pelem
+      in
+      let opamfile = OpamParser.FullPos.string txt filename in
+      let elements, lastpos =
+        List.fold_left nc_item ([], (1,1)) opamfile.file_contents
+      in
+      let str = String.concat "" (List.rev elements) in
+      let str =
+        let final = extract (pos_index lastpos) (String.length txt -1) in
+        str ^ final
+      in
+      let trim_end_whitespaces str =
+        let _rev_split seps str =
+          let length = String.length str in
+          let rec aux index beg acc =
+            if index = length -1 then acc else
+            if List.exists (fun sep -> str.[index] = sep) seps then
+              aux (index + 1)
+                (index +1)
+                (String.sub str index 1
+                 :: String.sub str beg (index  - beg)
+                 :: acc)
+            else
+              aux (index +1) beg acc
+          in
+          aux 0 0 []
+        in
+        let split_on_chars seps s =
+          let r = ref [] in
+          let j = ref (String.length s) in
+          for i = String.length s - 1 downto 0 do
+            if List.exists (fun sep -> String.get s i = sep) seps then begin
+              r := String.sub s (i + 1) (!j - i - 1) :: !r;
+              j := i
+            end
+          done;
+          String.sub s 0 !j :: !r
+        in
+        let empty_end str =
+          let length = String.length str in
+          if length = 0 then str else
+          let rec aux current =
+            if current = -1 then current else
+            if str.[current] = ' ' then aux (current -1)
+            else current
+          in
+          let index = aux (length -1) in
+          if index = -1 then "" else
+          if index = length -1 then str else
+            String.sub str 0 (index + 1)
+        in
+        str
+        |> split_on_chars [ '\n'; '\r' ]
+        |> List.map empty_end
+        |> List.filter (fun x -> x != "")
+        |> String.concat "\n"
```

Comment:
ping @dra27 for crlf

Reply:
It's no worse than the rest of the library does - we'll sort this out with #11's rebase for opam 2.2.